Add Application metrics Section in User Docs

sablumiah opened the pull request:
WHAT
Add section on how to create application metrics 

WHY
This will enable users to create an application metrics endpoint via instrumentation or using an exporter, and then create alerts based on those metrics using Prometheus

Commit:
Remove unused image files

Commit:
Add an image showing the deployment process

Commit:
Update the 'deploying an app.' walkthrough

This uses a much simpler example application, with
the minimum possible kubernetes configuration.
Some outdated values have been corrected (e.g. the
hostname at which the deployed application can be
accessed).

Commit:
Add a 'Cleaning Up' section to the user guide

This commit adds a brief explanation of how to
remove unneeded cluster resources that the user
created while working through sections of the
guide.

The 'Cleaning Up' section is added as the last
part of the 'Deploying an app' part of the user
docs.

Commit:
Add an example to decode the ECR secrets

As I've had to run this a few times and have to keep googling it...
Note: the example text decodes to `Aladdin:open sesame` and is not
an actual secret

Commit:
Repeat the `kubectl get secret` command

This may save the reader the hassle of following
the link to the ECR section and then coming back.

Commit:
Tweak titles: clarify references to 'the cluster'

The user isn't configuring the cluster, just their
namespace. Similarly, we don't need to specify
that we're deploying the application 'to the
cluster'

sablumiah committed:
Add applcaition metrics section

sablumiah committed:
Add applcaition metrics section

sablumiah committed:
add command to check networkpolices

sablumiah committed:
add command to check networkpolices

digitalronin committed:
Merge pull request #115 from ministryofjustice/davids-revisions

Update the 'deploying an app.' walkthrough

Commit:
Add holding page for multicontainer app. deploy

* Rename pages 002-005 to make room for a new 002
* Retitle 001 to 'Deploying a hello world app'
* Add a placeholder page for multicontainer deploy

Commit:
'Deploying an app' -> 'Deploying applications'

I think this is a more accurate title for this
section of the document

Commit:
Change absolute links to relative

This makes things less confusing when using a
development jekyll server, while editing the
document. It also protects against the 
documentation repository being moved, in the 
future.

Commit:
WiP: Add multi-container app. deployment walkthru

Commit:
WiP: Add multi-container app. deployment walkthru

Commit:
WiP: Add multi-container app. deployment walkthru

Commit:
WiP: Add multi-container app. deployment walkthru

Commit:
Git ignore .bundle directory

Commit:
Add a section on deleting namespace resources

Commit:
Move RDS section to the top

This requires approval from the cloud platform team, and takes a
while to apply once approved. So, it makes sense to tell the user
to do this first.

Commit:
Replace '-n' with '--namespace'

Commit:
WiP: Describe the kubernetes configuration

Commit:
WiP: Describe the kubernetes configuration

Commit:
WiP: Describe the kubernetes configuration

Commit:
WiP: Describe the kubernetes configuration

jasonBirchall commented on 03-monitoring-an-app/002-how-to-create-alarms.md:
```diff
@@ -93,6 +93,111 @@ kubectl -n <namespace> describe prometheusrules prometheus-custom-rules-<applica
 ## PrometheusRule examples
 If you're struggling for ideas on how and which alerts to setup, please see some examples [here](https://github.com/ministryofjustice/cloud-platform-infrastructure/blob/master/terraform/cloud-platform-components/resources/prometheusrule-examples/application-alerts.yaml).
 
+## Application Metrics
```
Due to the volume of this entry, I would suggest this needs its own document.

jasonBirchall reviewed (changes requested):
Can we move this to its own document?

jennyd commented:
I think it's worth including a mention of an important consideration: if you're using a pre-forking web server (like unicorn or puma for Ruby, or gunicorn for Python) and have it configured to use multiple processes, then you need to use a Prometheus client library which supports exporting metrics from multiple processes. Not all the official clients do that - for example https://github.com/prometheus/client_ruby/issues/9 has been open for a long time, although https://github.com/prometheus/client_ruby/pull/95 looks promising so :crossed_fingers:

If you don't use a library which supports this, then requests to `/metrics` could be served by any of the processes, which would mean Prometheus sees inconsistent data on each scrape.

Commit:
Correct 'fork' to 'branch', in cleanup section

Our policy for the environments repo is for developers to create
branches with the changes they want, not forks.

digitalronin committed:
Merge pull request #117 from ministryofjustice/cleanup

Add a 'Cleaning Up' section to the user guide

Commit:
Make the document headings consistent

* Add 'Step 4' (since we use 'Step 1-3' earlier)
* Use imperative tense, like the other steps

Commit:
Explain `kubectl delete --filename` to delete

Describe the process of using `kubectl delete`
with a directory full of yaml files to delete all
named resources from the user's namespace using a
single command.

digitalronin committed:
Merge pull request #122 from ministryofjustice/tweak-cleanup

Explain `kubectl delete --filename` to delete

digitalronin committed:
Merge pull request #121 from ministryofjustice/tweak-helloworld-deploy

Make the document headings consistent

Commit:
Correct the ordering of build/tag/push steps

Commit:
Correct the ordering of build/tag/push steps

Commit:
Mention demo. app. makefile, for build/tag/push

Commit:
Encourage the user to view the demo app README

Commit:
Remind user to check docker image names in k8s yml

If the user tags their docker images with a
different version number (using the makefile in
the demo app), they will need to update the image
reference in the kubernetes deployment yaml files.

Commit:
Warn the user to expect hostname conflicts

It's quite likely that someone will already have
deployed an instance of the demo app at

https://multi-container-demo.apps.cloud-platform-live-0.k8s.integration.dsd.io/

In that case, creating the ingress on the cluster
will fail. This warns the user to expect that
error.

Commit:
Add deployment instructions

Commit:
Use simpler command for base64 decode example

Commit:
Remove unnecessary backticks around base64 example

digitalronin committed:
Merge pull request #118 from ministryofjustice/add-secrets-decode-example

Add an example showing use of base64 to decode ECR secrets

Commit:
Add description of https headers for probes

For rails applications with force_ssl = true, probes need extra
http headers. Adding this explanation here will help developers
avoid problems with failing probes, without our help.

Commit:
Remove the exact redirect status code (301)

Rails 5 sends a 307, not a 301. We don't really care which
redirect we're avoiding.

digitalronin committed:
Merge pull request #124 from ministryofjustice/https-probes

Add description of using https headers for http liveness/readiness probes, where the application only responds to https.

Commit:
Amend instructions to git clone via https, not ssh

If users are not in the MoJ/webops github team, they won't be able
to clone the helloworld demo app. via SSH.

digitalronin committed:
Merge pull request #126 from ministryofjustice/tweak-helloworld

Amend instructions to git clone via https, not ssh

Commit:
Add instructions on how to contribute to the guide

People who are familiar with github pages will know how to go from

https://ministryofjustice.github.io/cloud-platform-user-docs

...to

https://github.com/ministryofjustice/cloud-platform-user-docs

For those who don't, this change provides instructions on how to
find the source repo and contribute to the guide.

digitalronin committed:
Merge pull request #123 from ministryofjustice/deploy-multi-container-app

Add a guide for deploying a multi-container app

digitalronin committed:
Merge pull request #127 from ministryofjustice/add-link-to-repo

Add instructions on how to contribute to the guide

Commit:
Add a deprecation warning for the live-0 cluster

This is a temporary measure to forewarn teams that live services
will be hosted on the live-1 cluster (although it's not quite
ready yet).

AntonyBishop committed:
Removed extra "s" from #ask-cloud-platform slack channel name.

Commit:
Correct 'availability zone' to 'region'

Commit:
Merge branch 'live0-deprecation-notice' of github.com:ministryofjustice/cloud-platform-user-docs into live0-deprecation-notice

digitalronin committed:
Merge pull request #130 from ministryofjustice/live0-deprecation-notice

Add a deprecation warning for the live-0 cluster

sablumiah committed:
Move applcaition metrics section to its own page

sablumiah committed:
Merge branch 'applicationmetrics' of github.com:ministryofjustice/cloud-platform-user-docs into applicationmetrics
